### PR TITLE
Add BBH executable

### PR DESCRIPTION
**Pull request by geoffrey4444**
## Proposed changes

This PR adds an executable for an executable that evolves binary black holes. Depends on #3460 and #3461; the new code is only in the commit labeled "Add exec to evolve binary black holes."

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by geoffrey4444** [comment]
I tested that I can start a BBH evolution using numeric initial data with the code in this PR. It's waiting for some smaller PRs, but in case there are bigger design changes requested, I thought I'd go ahead and submit this now

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHoleFwd.hpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+namespace gr {
+namespace Solutions {
+struct KerrSchild;
+}  // namespace Solutions
+}  // namespace gr
+namespace evolution {
+struct NumericInitialData;
+}  // namespace evolution
+
+template <size_t VolumeDim, typename InitialData, typename BoundaryConditions>
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHoleFwd.hpp]: you don't need any of these template parameters (or the forward includes)

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+template <typename Metavariables>
+class CProxy_GlobalCache;
+}  // namespace Parallel
+/// \endcond
+
+// Second template parameter specifies the source of the initial data, which
+// could be an analytic solution, analytic data, or imported numerical data.
+// Third template parameter specifies the analytic solution used when imposing
+// dirichlet boundary conditions or against which to compute error norms.
+template <size_t VolumeDim, typename InitialData, typename BoundaryConditions>
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp]: get rid of the template parameters.  VolumeDim = 3, InitialData = evolution::NumericInitialData and you shouldn't use BoundaryConditions

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+struct EvolutionMetavars {
+  static constexpr size_t volume_dim = VolumeDim;
+  static constexpr bool use_damped_harmonic_rollon = false;
+  using initial_data = InitialData;
+  using frame = Frame::Inertial;
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp]: Does anything outside this file use this type alias? If not, I would get rid of it

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+  // 2nd or 3rd order piecewise polynomial functions of time using
+  // `read_spec_piecewise_polynomial()`
+  static constexpr bool override_functions_of_time = true;
+
+  using normal_dot_numerical_flux = Tags::NumericalFlux<
+      GeneralizedHarmonic::UpwindPenaltyCorrection<volume_dim>>;
+
+  using time_stepper_tag = Tags::TimeStepper<
+      std::conditional_t<local_time_stepping, LtsTimeStepper, TimeStepper>>;
+  using analytic_solution_fields = typename system::variables_tag::tags_list;
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp]: get rid of

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+
+  using initialize_initial_data_dependent_quantities_actions = tmpl::list<
+      GeneralizedHarmonic::gauges::Actions::InitializeDampedHarmonic<
+          volume_dim, use_damped_harmonic_rollon>,
+      GeneralizedHarmonic::Actions::InitializeConstraints<volume_dim>,
+      Parallel::Actions::TerminatePhase>;
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& /*p*/) noexcept {}
+  using analytic_solution_tag = Tags::AnalyticSolution<BoundaryConditions>;
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp]: get rid of

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+  };
+
+  using observe_fields = tmpl::append<
+      tmpl::push_back<
+          analytic_solution_fields, gr::Tags::Lapse<DataVector>,
+          ::Tags::PointwiseL2Norm<
+              GeneralizedHarmonic::Tags::GaugeConstraint<volume_dim, frame>>,
+          ::Tags::PointwiseL2Norm<GeneralizedHarmonic::Tags::
+                                      ThreeIndexConstraint<volume_dim, frame>>>,
+      std::conditional_t<volume_dim == 3,
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp]: get rid of conditional

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+    Register,
+    LoadBalancing,
+    WriteCheckpoint,
+    Evolve,
+    Exit
+  };
+
+  static std::string phase_name(Phase phase) noexcept {
+    if (phase == Phase::LoadBalancing) {
+      return "LoadBalancing";
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp:242]: checkpointing should be in this list

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+    }
+    ERROR(
+        "Passed phase that should not be used in input file. Integer "
+        "corresponding to phase is: "
+        << static_cast<int>(phase));
+  }
+
+  using phase_changes =
+      tmpl::list<PhaseControl::Registrars::VisitAndReturn<EvolutionMetavars,
+                                                          Phase::LoadBalancing>,
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp:254]: checkpointing should be in this list

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+          cache_proxy) noexcept {
+    const auto next_phase = PhaseControl::arbitrate_phase_change<phase_changes>(
+        phase_change_decision_data, current_phase,
+        *(cache_proxy.ckLocalBranch()));
+    if (next_phase.has_value()) {
+      return next_phase.value();
+    }
+    switch (current_phase) {
+      case Phase::Initialization:
+        return evolution::is_numeric_initial_data_v<initial_data>
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp]: get rid of conditional

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+                                 volume_dim, Frame::Inertial>>>>>>;
+
+  using initialization_actions = tmpl::list<
+      Actions::SetupDataBox,
+      Initialization::Actions::TimeAndTimeStep<EvolutionMetavars>,
+      evolution::dg::Initialization::Domain<volume_dim,
+                                            override_functions_of_time>,
+      Initialization::Actions::NonconservativeSystem<system>,
+      std::conditional_t<
+          evolution::is_numeric_initial_data_v<initial_data>, tmpl::list<>,
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp]: get rid of conditional

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 11]
```diff
@@ -0,0 +1,444 @@
...
+          evolution::Initialization::Actions::SetVariables<
+              ::domain::Tags::Coordinates<volume_dim, Frame::Logical>>>,
+      Initialization::Actions::TimeStepperHistory<EvolutionMetavars>,
+      GeneralizedHarmonic::Actions::InitializeGhAnd3Plus1Variables<volume_dim>,
+      Initialization::Actions::AddComputeTags<tmpl::push_back<
+          StepChoosers::step_chooser_compute_tags<EvolutionMetavars>,
+          evolution::Tags::AnalyticCompute<volume_dim, analytic_solution_tag,
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp]: get rid of

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+      ::evolution::dg::Initialization::Mortars<volume_dim, system>,
+      evolution::Actions::InitializeRunEventsAndDenseTriggers,
+      Initialization::Actions::RemoveOptionsAndTerminatePhase>;
+
+  using gh_dg_element_array = DgElementArray<
+      EvolutionMetavars,
+      tmpl::flatten<tmpl::list<
+          Parallel::PhaseActions<Phase, Phase::Initialization,
+                                 initialization_actions>,
+          tmpl::conditional_t<
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp]: get rid of conditional

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp [hunk 13]
```diff
@@ -0,0 +1,444 @@
...
+  struct registration_list {
+    using type = std::conditional_t<
+        std::is_same_v<ParallelComponent, gh_dg_element_array>,
+        dg_registration_list, tmpl::list<>>;
+  };
+
+  using component_list = tmpl::flatten<tmpl::list<
+      observers::Observer<EvolutionMetavars>,
+      observers::ObserverWriter<EvolutionMetavars>,
+      std::conditional_t<evolution::is_numeric_initial_data_v<initial_data>,
```
Comment (kidder) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveBinaryBlackHole.hpp]: get rid of conditional

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+        FunctionOfTimeName: "RotationAngle"
+      SizeMap:
+        InitialValues: [0.0, 0.0]
+        InitialVelocities: [0.0, 0.0]
+        InitialAccelerations: [0.0, 0.0]
+        FunctionOfTimeNames: ["LambdaFactorA0", "LambdaFactorB0"]
+
+EventsAndDenseTriggers:
+
+AnalyticSolution:
```
Comment (kidder) [inline tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml]: shouldn't need this

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by geoffrey4444** [comment]
@kidder thanks for your help! I rebased and made your suggested changes, including removing the analytic solution

**Comment by kidder** [comment]
rebase and squash

**Comment by kidder** [comment]
@geoffrey4444 this is not compiling...

**Comment by geoffrey4444** [comment]
@kidder Thanks for letting me know...when I squashed, I missed that ObserveNorms is now needs ::Tags::Time as an additional template parameter. I only had a few minutes this morning, so I thought I'd try getting away without compiling locally. 🤦‍♂️ 

I just pushed a fix. I made sure this time that it builds locally for me, and I'll keep my eye on GitHub to make sure everything looks good

**Comment by geoffrey4444** [comment]
The clang-8 test failure is a timeout in `InputFiles.Poisson.ProductOfSinusoids1D.yaml.execute_check_output`, which I didn't touch in this PR

**Review by kidder — APPROVED** [review]

**Pull request merged**